Test/dafny0/NoMoreAssume2Less2.dfy

```diff
@@ -0,0 +1,199 @@
+// RUN: %dafny "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// This file tests various ways of writing the equivalent of "assume false;" or "assert false;" inside a loop.
+// It used to be that Dafny's translation of such statements into Boogie could cause Boogie to cut off the
+// loop back edge, which would turn a loop into something that wasn't a loop. In particular, it would cause
+// the expected "havoc loop_targets;" statement never to be generated, whose dramatic impact on what gets
+// verified would be both mysterious and surprising to most Dafny users.
+//
+// The equally mysterious workaround for this problem used to be for Dafny users to replace their "assume false;"
+// with something like "assume 2 < 2;", which Boogie wouldn't recognize in the same way.
+//
+// The current translation to Boogie makes sure that no user-supplied expression turns into exactly an
+// "assume false;" or "assert false;". This means that Dafny users can use a statement like "assume false;"
+// without chopping off back edges.
+//
+// It is still possible (in situations that would be rare in practice) to get some initially-surprising behavior
+// from Boogie, because it is possible to introduce a "false" condition that will be detected by the control-flow
+// handling in Boogie's abstract interpreter. Method Conjuncts4 below shows such an example.
+
+method M0(n: nat) {
+  var i := 0;
+  while i < n {
+    i := i + 1;
+  }
+  assert n == 0; // error
+}
+
+method M1(n: nat) {
+  var i := 0;
+  while i < n {
+    assume false;
+    i := i + 1;
+  }
+  assert n == 0; // error
+}
+
+method M2(n: nat) {
```

Comment:
What happens to a program like

```dafny
method TurnedOff() {
  assert false;
  assert 1 == 2;
}
```

Does the change in this PR cause a verification error to appear for `assert 1 == 2` ?

Reply:
No, there's no change for this example. This PR just makes sure that no `assume false` causes a loop back edge to disappear.

Reply:
I ran this branch on the program I pasted above and it resulted in:

```
Prover error: line 3802 column 30: model is not available

/Users/rwillems/Documents/Notes/Temp/d.dfy(1,7): Verification inconclusive (Impl$$_module.__default.TurnedOff)

Dafny program verifier finished with 0 verified, 0 errors, 1 inconclusive
```

Dafny program:
```
method TurnedOff() {
  assert false;
  assert 1 == 2;
}
```

Reply:
I don't know what you're seeing, and I don't understand how it would be connected to this PR. Before this PR, the `TurnedOff()` method would have produced the following Boogie:

```
...
assert false;
...
assert LitInt(1) == LitInt(2);
```

After this PR, it instead produces

```
...
assert Lit(false);
...
assert LitInt(1) == LitInt(2);
```

Are you using some special switches to Boogie (for example, from the error message above, it seems Boogie is looking for a model)?